dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/LocaleController.java

```diff
@@ -71,4 +101,96 @@
+    {
+        response.setHeader( ContextUtils.HEADER_CACHE_CONTROL, CacheControl.noCache().cachePrivate().getHeaderValue() );
+        I18nLocale locale = localeService.getI18nLocaleByUid( uid );
+
+        if ( locale == null )
+        {
+            throw new WebMessageException( WebMessageUtils.notFound( "Cannot find Locale with uid: " + uid ) );
+        }
+
+        return locale;
+    }
+
+    @RequestMapping( value = "/languages", method = RequestMethod.GET, produces = "application/json" )
+    public @ResponseBody Map<String, String> getAvailableLanguages()
+    {
+        return localeService.getAvailableLanguages();
+    }
+
+    @RequestMapping( value = "/countries", method = RequestMethod.GET, produces = "application/json" )
+    public @ResponseBody Map<String, String> getAvailableCountries()
+    {
+        return localeService.getAvailableCountries();
+    }
+
+    @PreAuthorize( "hasRole('ALL') or hasRole('F_LOCALE_ADD')" )
+    @RequestMapping( method = RequestMethod.POST )
+    @ResponseStatus( value = HttpStatus.OK )
+    public void addLocale( @RequestParam String country, @RequestParam String language,
+        HttpServletRequest request, HttpServletResponse response ) throws WebMessageException
+    {
+        if ( StringUtils.isEmpty( country ) || StringUtils.isEmpty( language ) )
+        {
+            throw new WebMessageException( WebMessageUtils.conflict( "Invalid country or language code." ) );
+        }
+
+        String localeCode = LocaleUtils.getLocaleString( language, country, null );
+
+        Locale locale = LocaleUtils.getLocale( localeCode );
+
+        if ( locale != null )
+        {
+            I18nLocale i18nLocale = localeService.getI18nLocale( locale );
+
+            if ( i18nLocale != null )
+            {
+                throw new WebMessageException( WebMessageUtils.conflict( "Locale code existed." ) );
+            }
+        }
+
+        I18nLocale i18nLocale = localeService.addI18nLocale( language, country );
+
+        WebMessage webMessage = WebMessageUtils.created( "Locale created successfully" );
+
+        response.setHeader( ContextUtils.HEADER_LOCATION, contextService.getApiPath() + "/locales/" + i18nLocale.getUid() );
+
+        webMessageService.send( webMessage, response, request );
+    }
+
+    @PreAuthorize( "hasRole('ALL') or hasRole('F_LOCALE_DELETE')" )
+    @RequestMapping( path = "/{uid}" ,method = RequestMethod.DELETE )
+    @ResponseStatus( value = HttpStatus.NO_CONTENT )
+    public void delete( @PathVariable String uid ) throws WebMessageException
+    {
+        try
+        {
+            I18nLocale i18nLocale = localeService.getI18nLocaleByUid( uid );
+
+            if ( i18nLocale == null )
+            {
+                throw new WebMessageException( WebMessageUtils.conflict( "Cannot find Locale with uid " + uid ) );
+            }
+
+            localeService.deleteI18nLocale( i18nLocale );
+        }
+        catch ( DeleteNotAllowedException ex )
```

Comment:
Should we catch this specifically there? I.e. is this really a cross-cutting concern which can happen all places? I.e do we already handle it in the controller advice?

Reply:
@larshelge you are right, it is handled in ControllerAdvice. Removed it now.